contrib/babelfishpg_tsql/src/pltsql_utils.c

```diff
@@ -3093,3 +3093,137 @@ is_valid_func_ownership_chain(void *expr, Oid objectOwnerId)
 	}
 	return (OidIsValid(immediate_parent_func) && (get_func_owner(immediate_parent_func) == objectOwnerId));
 }
+
+/*
+ * restrict_alter_owner_stmt
+ * Blocks ALTER OWNER statements from PG dialect on TSQL objects.
+ * Allows ALTER OWNER for specific shipped objects only when new owner is sysadmin.
+ */
+void
+restrict_alter_owner_stmt(AlterOwnerStmt *stmt)
+{
+    char *schema_name = NULL;
+    char *object_name = NULL;
+    Oid schema_oid = InvalidOid;
+
+    /* Extract schema and object name based on object type */
+    switch (stmt->objectType)
+    {
+        case OBJECT_FUNCTION:
+        case OBJECT_PROCEDURE:
+		{
+			ObjectWithArgs *owa = (ObjectWithArgs *) stmt->object;
+			if (list_length(owa->objname) > 1)
+			{
+				schema_name = strVal(linitial(owa->objname));
+				object_name = strVal(lsecond(owa->objname));
+			}
+			else if (list_length(owa->objname) == 1)
+			{
+				object_name = strVal(linitial(owa->objname));
+			}
+			break;
+		}
+        case OBJECT_SCHEMA:
+		{
+			String *str = (String *) stmt->object;
+			schema_name = strVal(str);
+			break;
+		}
+		case OBJECT_TYPE:
+		{
+			List *names = (List *) stmt->object;
+			if (names && list_length(names) > 1)
+				schema_name = strVal(linitial(names));
+			break;
+		}
+        default:
+            return; /* Allow other object types */
+    }
+
+    /* Get schema OID if we have schema name */
+    if (schema_name)
+        schema_oid = get_namespace_oid(schema_name, true);
+    else if (OidIsValid(schema_oid))
+        schema_name = get_namespace_name(schema_oid);
+
+    /* Check if it's a Babelfish schema */
+    if (schema_name && physical_schema_name_exists(schema_name))
+    {
+        /* Check if this is a shipped function being changed to sysadmin */
+        if (object_name && stmt->objectType == OBJECT_FUNCTION &&
+            is_ms_shipped((char*)object_name, OBJECT_TYPE_TSQL_SCALAR_FUNCTION, schema_oid))
```

Comment:
Can you please put a comment on why we need this special handling for ms shipped objects with some examples?